lib/commands/context.js

```diff
@@ -237,6 +248,155 @@ extensions.mobileGetContexts = async function mobileGetContexts () {
+    keyId = newIds[0] || null;
+  }
+
+  if (!util.hasValue(this.curContext)) {
+    log.debug('We do not appear to have window set yet, ignoring');
+    return;
+  }
+
+  const [curAppIdKey, curPageIdKey] = this.curContext.split('.');
+
+  if (curAppIdKey !== appIdKey) {
+    log.debug('Page change not referring to currently selected app, ignoring.');
+    return;
+  }
+
+  let newPage = null;
+  if (newPages.length) {
+    newPage = _.last(newPages);
+    log.debug(`We have new pages, selecting page '${newPage}'`);
+  } else if (!_.includes(newIds, curPageIdKey)) {
+    log.debug('New page listing from remote debugger does not contain ' +
+                 'current window; assuming it is closed');
+    if (!util.hasValue(keyId)) {
+      log.error('Do not have our current window anymore, and there ' +
+                   'are not any more to load! Doing nothing...');
+      this.setCurrentUrl(undefined);
+      return;
+    }
+
+    log.debug(`Debugger already selected page '${keyId}', ` +
+                 `confirming that choice.`);
+    this.curContext = `${appIdKey}.${keyId}`;
+    newPage = keyId;
+  } else {
+    // at this point, there are no new pages, and the current page still exists
+    log.debug('Checking if page needs to load');
+    // If a window navigates to an anchor it doesn't always fire a page
+    // callback event. Let's check if we wound up in such a situation.
+    const needsPageLoad = (() => {
+      // need to map the page ids to context ids
+      const contextArray = _.map(pageArray, (page) => `${appIdKey}.${page.id}`);
+      // check if the current context exists in both our recorded contexts,
+      // and the page array
+      return !_.isEqual(_.find(this.contexts, this.curContext), _.find(contextArray, this.curContext));
+    })();
+
+    if (needsPageLoad) {
+      log.debug('Page load needed. Loading...');
+      await this.remote.pageLoad();
+    }
+
+    log.debug('New page listing is same as old, doing nothing');
+  }
+
+  // make sure that the page listing isn't indicating a redirect
+  if (util.hasValue(this.curContext)) {
+    let currentPageId = parseInt(_.last(this.curContext.split('.')), 10);
+    let page = _.find(pageArray, (p) => parseInt(p.id, 10) === currentPageId);
+    if (page && page.url !== this.getCurrentUrl()) {
+      log.debug(`Redirected from '${this.getCurrentUrl()}' to '${page.url}'`);
+      this.setCurrentUrl(page.url);
+    }
+  }
+
+  if (util.hasValue(newPage)) {
+    this.selectingNewPage = true;
+    const oldContext = this.curContext;
+    try {
+      this.curContext = `${appIdKey}.${newPage}`;
+      // do not wait, as this can take a long time, and the response is not necessary
+      this.remote.selectPage(appIdKey, parseInt(newPage, 10))
+        .catch((err) => {
+          log.warn(`Failed to select page: ${err.message}`);
+          this.curContext = oldContext;
+        });
+    } finally {
+      this.selectingNewPage = false;
```

Comment:
selectPage is never going to fail here, as it is a promise, so I don't quite understand why do we need `finally` block? Do we want to add `.finally` clause to `selectPage` instead (it has to be a bluebird-compatible promise to support that)?

Reply:
Good point. Thanks!